src/Core/Occurrence.hs

```diff
@@ -2,126 +2,182 @@
 module Core.Occurrence
   ( OccursVar(..)
   , Occurs(..)
-  , tagOccurrence, doesItOccur
+  , Occurrence(..)
+  , tagOccursVar, tagOccursMap
+  , tagOccurStmt, tagOccurTerm
+  , doesItOccur
   ) where
 
 
 import Core.Core as C
 
-import qualified Data.Map.Merge.Strict as Map
-import qualified Data.Map.Strict as Map
+import qualified Data.VarMap as VarMap
 import Data.VarSet (IsVar(..))
+import Data.Semigroup
 import Data.Generics
-import Data.Maybe
-import Data.List
 import Data.Triple
 
-import Control.Monad.Infer
-import Control.Arrow
-
-import Syntax (Var(..), Resolved)
 import Pretty
 
-type OccAtom a = C.Atom (OccursVar a)
-type OccTerm a = C.Term (OccursVar a)
-type OccType a = C.Type (OccursVar a)
-type OccStmt a = C.Stmt (OccursVar a)
+data Occurrence = Dead | Once | OnceLambda | Multi | MultiLambda
```

Comment:
Have you been reading _Secrets of the GHC Inliner_?

Reply:
Yeah. It doesn't include any of the loop breaker information (yet), but this allows us to reuse it for lifting.